Unix/http/httpclient.c

```diff
@@ -1843,6 +1899,266 @@ static Page* _CreateHttpHeader(
+
+
+    /* First delivery. pAuthType. We convert the string into an enum */
+    
+    method = AUTH_METHOD_NONE;
+    if (userCredentials.authenticationType)
+    {
+        method = AUTH_METHOD_UNSUPPORTED;
+    }
+
+    if (Tcscasecmp(userCredentials.authenticationType, AUTH_NAME_BASIC) == 0)
+    {
+        method = AUTH_METHOD_BASIC;
+    }
+   
+    if (Tcscasecmp(userCredentials.authenticationType, AUTH_NAME_NEGOTIATE) == 0) 
+    {
+        method = AUTH_METHOD_NEGOTIATE;
+    }
+    
+    if (Tcscasecmp(userCredentials.authenticationType, AUTH_NAME_NEGOTIATE_WITH_CREDS) == 0) 
+    {
+        method = AUTH_METHOD_NEGOTIATE_WITH_CREDS;
+    }
+    
+    if (Tcscasecmp(userCredentials.authenticationType,  AUTH_NAME_KERBEROS) == 0) 
+    {
+        method = AUTH_METHOD_KERBEROS;
+    }
+
+    // We only output it if it is wanted.
+
+    if (pAuthType) {
+        *pAuthType = method;
+    }
+
+    if (method == AUTH_METHOD_UNSUPPORTED)
+    {
+        // Log here
+        LOGE2((ZT("_UnpackDestinationOptions: Authorisation type (%s) is not supported."), userCredentials.authenticationType));
+        result = MI_RESULT_INVALID_PARAMETER;
+        goto Done;
+    }         
+
+
+    if (userCredentials.credentials.usernamePassword.username) {
+        username_len = Tcslen(userCredentials.credentials.usernamePassword.username);
+        username = (char*) PAL_Malloc(username_len+1);
+        if (username == NULL)
+        {
+            result = MI_RESULT_SERVER_LIMITS_EXCEEDED;
+            goto Done;
+        }
+        memcpy(username, userCredentials.credentials.usernamePassword.username, username_len+1);
+        username[username_len] = 0;
+    }
+    else {
+        if (method == AUTH_METHOD_BASIC) {
+            LOGE2((ZT("_UnpackDestinationOptions: Authorisation type Basic requires username.")));
+            return MI_RESULT_INVALID_PARAMETER;
+        }
+    }
+
+    if (pUsername) {
+        *pUsername = username;
+    }
+
+    /* We need to allocate a buffer for the password.
+     * Get length of it, allocate and then retrieve it.
+     */
+
+    if (MI_DestinationOptions_GetCredentialsPasswordAt(pDestOptions, 0, (const MI_Char **)&optionName, NULL, 0, &password_len, NULL) != MI_RESULT_OK)
+    {
+        if (method == AUTH_METHOD_BASIC) {
+            LOGE2((ZT("_UnpackDestinationOptions: Authorisation type requires password.")));
+            return MI_RESULT_INVALID_PARAMETER;
+        }
+    }
+
+    if (password_len <= 0) {
+        if (method == AUTH_METHOD_BASIC) {
+            LOGE2((ZT("_UnpackDestinationOptions: Authorisation type requires password.")));
+            return MI_RESULT_INVALID_PARAMETER;
+        }
+    }
+    else 
+    {
+#if defined(CONFIG_ENABLE_WCHAR)
+        MI_Char *wide_password = (MI_Char*) PAL_Malloc(password_len*sizeof(MI_Char));
+
+        password = (char*) PAL_Malloc(password_len);
+        if (password == NULL)
+        {
+            return MI_RESULT_SERVER_LIMITS_EXCEEDED;
```

Comment:
leaking username and others? Actually, a bunch of these failure points should probably set error and goto Done;